audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,506 @@
+      // computer tests.
+      audit.define('attack-and-release', (task, should) => { 
+        let context = new OfflineAudioContext(2, testFrames, sampleRate);
+        // Attack and release are set to be larger than buffer size to ensure
+        // that the test success depends on proper noise gate memory between
+        // blocks.
+        const attackAndReleaseSteps = bufferSize + 128;
+        const attack = _toTime(attackAndReleaseSteps, sampleRate) * 1000;
+        const release = attack;
+        let firstDecreasingIndex = attackAndReleaseSteps;
+        const startOfDrop = _toTime(firstDecreasingIndex, sampleRate);
+        // The amount of time it takes for gain to move from a value above the
+        // threshold to a value beneath the threshold is set to be larger than
+        // the intended amount of attack and release steps so that a linear
+        // decrease can be verified for the attacking phase.
+        const gainChangeTime = _toTime(attackAndReleaseSteps, sampleRate); 
+        const endOfDrop = startOfDrop + gainChangeTime;
+        // The signal will rise some number of indexes later than
+        // startOfDrop to allow time to inspect the fully closed gate.
+        const startOfRise = endOfDrop + _toTime(bufferSize, sampleRate);
+        const endOfRise = startOfRise + gainChangeTime;
+        const gain = new GainNode(context, {gain:aboveThresholdGain});
+        gain.gain.linearRampToValueAtTime(aboveThresholdGain, startOfDrop);
+        gain.gain.linearRampToValueAtTime(belowThresholdGain, endOfDrop);
+        gain.gain.linearRampToValueAtTime(belowThresholdGain, startOfRise);
+        gain.gain.linearRampToValueAtTime(aboveThresholdGain, endOfRise);
+
+        let noisegate = new NoiseGate(
+            context, {bufferSize, attack, release, threshold, alpha});
+        let delay = new DelayNode(
+            context, {delayTime : _toTime(scriptProcessorLatency, sampleRate)});
+        let oscillator = new OscillatorNode(context);
+        let merger = new ChannelMergerNode(context, {numberOfInputs: 2});
+
+        oscillator.connect(gain).connect(noisegate.input);
+        gain.connect(delay);
+        noisegate.output.connect(merger, 0, 0);
+        delay.connect(merger, 0, 1);
+        merger.connect(context.destination);
+
+        // Drop the bass! 
+        oscillator.start();
+
+        context.startRendering().then((buffer) => {
+          const noiseGateOutput = buffer.getChannelData(0);
+          const originalOutput = buffer.getChannelData(1);
+
+          // Verify 6 phases of the output:
+          // 1. Latency Phase: all zeros until scriptProcessorLatency
+          // 2. Open Phase 1: unaffected until firstDecreasingIndex
+          // 3. Attacking Phase: decreasing over time until 0
+          // 4. Closed Phase: all zeros until firstReleaseIndex
+          // 5. Releasing Phase: increasing until identical to original output
+          // 6. Open phase 2: unaffected until the last sample
```

Comment:
I don't see how that would reduce the complexity. Wouldn't it just mean a lot of code duplication?